src/python/pants/engine/engine.py

```diff
@@ -109,10 +112,18 @@ def product_request(self, product, subjects):
 
     # Throw handling.
     # TODO: See https://github.com/pantsbuild/pants/issues/3912
-    throw_roots = tuple(root for root, state in result_items if type(state) is Throw)
-    if throw_roots:
-      cumulative_trace = '\n'.join(self._scheduler.trace())
-      raise ExecutionError('Received unexpected Throw state(s):\n{}'.format(cumulative_trace))
+    throw_root_states = tuple(state for root, state in result_items if type(state) is Throw)
+    if throw_root_states:
+      if self._include_trace_on_error:
+        cumulative_trace = '\n'.join(self._scheduler.trace())
+        raise ExecutionError('Received unexpected Throw state(s):\n{}'.format(cumulative_trace))
+
+      if len(throw_root_states) == 1:
+        raise throw_root_states[0].exc
+      else:
+        raise ExecutionError('Multiple exceptions encountered:\n  {}'
```

Comment:
Getting multiple failed roots here doesn't mean multiple failures, necessarily. Because a `Throw` in the graph with two "parent" nodes will fail both parents with the same `Throw`.

Deduping by the string value of the throw might work, but the real right way would be to find the lowest point in the graph, which is where the `Throw` originated (as trace does). Deduping is probably fine, but executing a different native method might be cleaner?